language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+# Puppet Resource Types
+
+This document describes puppet's Ruby DSL for defining resource types, aka [custom types](https://docs.puppet.com/guides/custom_types.html#types). Examples are taken from the `puppetlabs-aws` module and core puppet.
+
+The second part of the document describes issues that cause puppet resource types to leak across environments (lack of environment isolation).
+
+## Puppet::Type.newtype method
+
+Defines a new puppet resource type and registers the type using the specified symbolic name. For example, to define a resource type named `ec2_instance`: 
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  # definition goes here
+end
+```
+
+The `newtype` method creates a new class `Puppet::Type::Ec2_instance`, which subclasses `Puppet::Type`. The block passed to the `newtype` method defines methods and variables on the `Puppet::Type::Ec2_instance` class.
+
+## Puppet::Type::&lt;name&gt; methods
+
+Within the `Puppet::Type.newtype` block, the following DSL methods are often called, and are evaluated in the context of the  `Puppet::Type::<name>` class:
+
+### doc string
+
+Defines the documentation string for the type, which is output when running `puppet describe <type>`. The doc string is usually set via the `Puppet::Type` class variable, though there is also a `Puppet::Type.doc=` class method.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  @doc = 'A type representing an AWS VPC'
+end
+```
+
+### namevar
+
+The type must specify a parameter to be the name of the resource, aka namevar. The name must uniquely identity that resource. By default, a parameter named `name` will be the namevar. This behavior can be modified using `title_patterns` and/or explicitly declaring a parameter to be a namevar.
```

Comment:
I'll change that to `isnamevar` since that's the actual DSL method. It can also be set in the options hash, e.g. `namevar => true`.

This is a bit in the weeds, but technically the `Puppet::Type#name_var` will return the parameter name (Symbol) or names (Array&lt;Symbol&gt;) that determine the identity for that instance. For example, given:

```ruby
Puppet::Type.newtype(:foo) do
  newparam(:name) do
  end
end
```

Then `Puppet::Type::Foo#name_var` returns:

```
irb(main):011:0> Puppet::Type.type(:foo).new(:name => 'bar').name_var
=> :name
```
That logic will include any parameter whose `Puppet::Parameter.isnamevar?` is truthy or any parameter whose name is `:name`:

```
irb(main):003:0> Puppet::Type.type(:foo).attrclass(:name)
=> Puppet::Type::Foo::ParameterName
irb(main):004:0> Puppet::Type.type(:foo).attrclass(:name).isnamevar?
=> nil
irb(main):005:0> Puppet::Type.type(:foo).key_attribute_parameters
=> [Puppet::Type::Foo::ParameterName]
```

The middle one is surprising because the `isnamevar?` method returns falsey, but puppet effectively treats it as a namevar in [`key_attribute_parameters`](https://github.com/puppetlabs/puppet/blob/4.4.1/lib/puppet/type.rb#L374-L380).